### PR TITLE
fix(auth): add basic auth challenge header

### LINKED ISSUE #2337: Should BasicAuthenticator add WWW-Authenticate header?
### Summary
Routes protected by `BasicAuthenticator`s do not properly trigger browser or `URLSession` authentication prompts due to the lack of a `WWW-Authenticate` header.

### Steps to reproduce
1. Create a simple `BasicAuthenticator`, like this one:

```swift
struct BasicPathAuthenticator: BasicAuthenticator {
    enum Error: Swift.Error { case invalidRequest, invalidCredentials }
    
    func authenticate(basic: BasicAuthorization, for request: Request) -> EventLoopFuture<Void> {
        guard let username = request.parameters["user", as: String.self],
            let password = request.parameters["passwd", as: String.self] else {
                return request.eventLoop.makeFailedFuture(Error.invalidRequest)
        }
        
        guard basic.username == username, basic.password == password else {
            return request.eventLoop.makeFailedFuture(Error.invalidCredentials)
        }
        
        request.auth.login(request)
        return request.eventLoop.makeSucceededFuture(())
    }
}
```

using it like this:

```swift
let basicAuth = app.grouped(BasicPathAuthenticator())
basicAuth.on([.GET, .POST, .PUT, .PATCH, .DELETE], "basic-auth", ":user", ":passwd") { request -> EventLoopFuture<Response> in
    guard request.isAuthenticated else {
        return request.eventLoop.makeSucceededFuture(Response(status: .unauthorized))
    }
    
    return try reply(to: request).encodeResponse(for: request)
}
```

2. Make a request, notice no automatic prompting from browser, or `URLSession` automatically applying `URLCredential`s.

### Expected behavior

Responses returned using the `BasicAuthorization` should have the `WWW-Authenticate` header applied automatically so that clients perform automatic behaviors.

### Actual behavior

Header has to be added manually in the route's response

### Environment

* Vapor Framework version: 4.4.0
* Vapor Toolbox version: ?
* OS version: 10.15.4

### Discussion

Perhaps I'm misunderstanding the role of `Middleware` here, or the role of `Authorization` middleware, but this seems like a desirable behavior. I was unaware of the `WWW-Authenticate` header and how it triggered the automatic behavior, which is my own ignorance of backend development. I was rather surprised to see a difference in behavior between the Basic auth offered by vapor and the basic auth offered by, say, a web server, like Apache, which adds the header automatically.

### PR DESCRIPTION
**Pull request by dioKaratzas**
Reopens #3457 against `main` as requested by @0xTim (the original PR targeted `vapor-5`, which has since been merged into `main`). Rebased onto the latest `main`.

---

Adds a typed `WWW-Authenticate` header helper and updates Basic authentication middleware to include a Basic challenge on unauthorized responses.

Previously, routes protected by `BasicAuthenticator` could return `401 Unauthorized` without a `WWW-Authenticate` header, which prevents clients such as browsers and `URLSession` from recognizing the response as an authentication challenge.

This change adds a default Basic realm of `Vapor`, while allowing authenticators to customize the realm by overriding `realm`:

```swift
struct AdminAuthenticator: BasicAuthenticator {
    var realm: String { "Admin" }

    func authenticate(basic: BasicAuthorization, for request: Request) async throws {
        // Authenticate credentials.
    }
}
```

Unauthorized Basic auth responses now include:

```http
WWW-Authenticate: Basic realm="Vapor"
```

If a response or thrown abort already includes `WWW-Authenticate`, Vapor preserves the existing header.

Fixes #2337.

### CODE REVIEW

File: Sources/Vapor/Authentication/Authenticator.swift [hunk 1]
```diff
@@ -27,10 +28,38 @@ extension RequestAuthenticator {
...
+            if response.status == .unauthorized, response.headers[.wwwAuthenticate] == nil {
+                response.headers.wwwAuthenticate = self.authenticateHeaderValue
+            }
+            return response
+        } catch let error as any AbortError where error.status == .unauthorized {
+            var headers = error.headers
+            if headers[.wwwAuthenticate] == nil {
+                headers.wwwAuthenticate = self.authenticateHeaderValue
+            }
+            throw Abort(.unauthorized, headers: headers, reason: error.reason)
```
Comment (0xTim) [inline Sources/Vapor/Authentication/Authenticator.swift]: We probably want to try and keep the source/details etc from the original error as well (might need to do a check to see if it conforms to `DebuggableError`)

File: Sources/Vapor/Authentication/Authenticator.swift [hunk 2]
```diff
@@ -27,10 +28,38 @@ extension RequestAuthenticator {
...
+            let response = try await next.respond(to: request)
+            if response.status == .unauthorized, response.headers[.wwwAuthenticate] == nil {
+                response.headers.wwwAuthenticate = self.authenticateHeaderValue
+            }
+            return response
+        } catch let error as any AbortError where error.status == .unauthorized {
+            var headers = error.headers
+            if headers[.wwwAuthenticate] == nil {
+                headers.wwwAuthenticate = self.authenticateHeaderValue
+            }
```
Comment (0xTim) [inline Sources/Vapor/Authentication/Authenticator.swift]: Could we do the same change in the bearer middleware as well?

File: Sources/Vapor/HTTP/Headers/HTTPFields+WWWAuthenticate.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import Foundation
```
Comment (0xTim) [inline Sources/Vapor/HTTP/Headers/HTTPFields+WWWAuthenticate.swift]: See the recent work for `FoundationEssentials` - we don't want to pull in all of Foundation on Linux

File: Tests/VaporTests/AuthenticationTests.swift [hunk 4]
```diff
@@ -96,6 +98,36 @@ struct AuthenticationTests {
         }
     }
 
+    @Test("Test Basic Authenticator WWW Authenticate Header")
```
Comment (0xTim) [inline Tests/VaporTests/AuthenticationTests.swift:108]: This is a good start but I think we need to expand the test suite:
* what happens if the user has manually set the `www-authenticate` header? We don't want to overwrite that
* test for bearer `www-authenticate` as well
* test throwing custom errors with reasons etc
* test a throwing authenticator

File: Tests/VaporTests/HTTPHeaderTests.swift [hunk 5]
```diff
@@ -111,6 +111,20 @@ struct HTTPHeaderTests {
         #expect(serializer.serialize() == "foo; bar=\"baz\", qux=\"quuz\"")
     }
 
+    @Test("Test WWW-Authenticate Header")
+    func testWWWAuthenticate() throws {
```
Comment (0xTim) [inline Tests/VaporTests/HTTPHeaderTests.swift:119]: Can we also test an empty realm and test a realm with a backslash in?

**Review by 0xTim — CHANGES_REQUESTED** [review]
This looks good! Just a few edge cases to sort out

**Comment by dioKaratzas** [comment]
Thanks @0xTim, think that's all of them.

The catch block now pulls  `identifier`  and  `source`  off the error when it's a  `DebuggableError`. And if the error already has a  `WWW-Authenticate`  set we don't catch it at all now, so it reaches  `ErrorMiddleware`  exactly as thrown.

Bearer got the same treatment,  `realm`  and the challenge handling, sharing a helper with Basic.

Dropped the Foundation import entirely rather than going the FoundationEssentials route  `replacingOccurrences`  was the only thing needing it, so the escaping is plain stdlib now.

Tests cover bearer, throwing authenticators, custom reasons, challenges set manually (returned and thrown), non-401 and non-`AbortError`, and empty/backslash realms.

Couple of things I wasn't sure on:

I added  `charset="UTF-8"`  to the Basic challenge, which is a bit beyond what you asked. Came out of  [#2337](https://github.com/dioKaratzas/vapor/issues/2337) , we decode credentials as UTF-8 but never advertised it, so clients fall back to ISO-8859-1 and non-ASCII passwords fail even when they're correct. Both workarounds in that issue set it. Say the word and I'll drop it.

Chaining is a bit odd. Middleware is applied outermost-first, so the innermost authenticator stamps the challenge and everything above sees a header already there, a route with Basic + Bearer only ever advertises one. RFC 7235 allows multiple so appending is probably better, but I wrote a test for the current behaviour instead of changing it. Separate issue?

Last thing, you mentioned in  [#2337](https://github.com/dioKaratzas/vapor/issues/2337)  that ErrorMiddleware should own this. I put it on the authenticators because ErrorMiddleware only sees thrown errors, and the original report returns a 401 straight from the handler, so that case never reaches it. It also can't know which scheme a route uses. Can move it if you still prefer that.

File: Sources/Vapor/HTTP/Headers/HTTPFields+WWWAuthenticate.swift [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    public struct WWWAuthenticate: ExpressibleByStringLiteral, Equatable, Sendable {
+        /// The serialized header value.
+        public let value: String
+
+        /// Creates a `WWW-Authenticate` header from a serialized header value.
+        public init(value: String) {
+            self.value = value
+        }
+
+        public init(stringLiteral value: String) {
```
Comment (0xTim) [inline Sources/Vapor/HTTP/Headers/HTTPFields+WWWAuthenticate.swift:17]: Missing DocC comment

**Review by 0xTim — CHANGES_REQUESTED** [review]
Thanks for this, this looks good! One minor nit and we're good. I'm going to hold off merging until #3467 is in as that should fix CI so we can make sure the tests pass

File: Sources/Vapor/HTTP/Headers/HTTPFields+WWWAuthenticate.swift (hunk 6 above)
Comment (dioKaratzas) [reply Sources/Vapor/HTTP/Headers/HTTPFields+WWWAuthenticate.swift:17]: Ready!
